src/wallet/wallet2.cpp

```diff
@@ -855,6 +855,46 @@ crypto::public_key wallet2::get_subaddress_spend_public_key(const cryptonote::su
   return D;
 }
 //----------------------------------------------------------------------------------------------------
+std::vector<crypto::public_key> wallet2::get_subaddress_spend_public_keys(uint32_t account, uint32_t begin, uint32_t end) const
+{
+  CHECK_AND_ASSERT_THROW_MES(begin <= end, "begin > end");
+
+  const cryptonote::account_keys& keys = m_account.get_keys();
+  std::vector<crypto::public_key> pkeys;
+  pkeys.reserve(end - begin);
+  cryptonote::subaddress_index index = {account, begin};
+
+  ge_p3 p3;
+  ge_cached cached;
+  CHECK_AND_ASSERT_THROW_MES(ge_frombytes_vartime(&p3, (const unsigned char*)keys.m_account_address.m_spend_public_key.data) == 0,
+      "ge_frombytes_vartime failed to convert spend public key");
+  ge_p3_to_cached(&cached, &p3);
+
+  for (uint32_t idx = begin; idx < end; ++idx)
+  {
+    index.minor = idx;
+    if (index.is_zero())
+    {
+      pkeys.push_back(keys.m_account_address.m_spend_public_key);
+      continue;
+    }
+    const crypto::secret_key m = cryptonote::get_subaddress_secret_key(keys.m_view_secret_key, index);
+
+    // M = m*G
+    ge_scalarmult_base(&p3, (const unsigned char*)m.data);
```

Comment:
Why is this aliasing `cryptonote::get_subaddress_public_key` ? If that implementation is slower, can it be updated? Also, this should have always been in a library function outside of wallet 2 that takes the base key.

Reply:
The function `cryptonote::get_subaddress_public_key` doesn't exist, and also isn't relevant here because we want the `ge_p3` part, not `crypto::public_key`. But I agree, probably it's better to have a function like this under `libcryptonote_basic`:

```c++
std::unordered_map<cryptonote::subaddress_index, crypto::public_key> cryptonote::get_subaddress_spend_public_keys(const crypto::public_key &spend_key, const std::unordered_set<cryptonote::subaddress_index> &indices)
```

so that wallet2 doesn't need to deal with `ge_*`.

Reply:
The speedup is from shedding unnecessary conversions, which doesn't apply when calculating one at a time (unless you want to start keeping data in low level types in wallet2).

Reply:
https://github.com/monero-project/monero/blob/master/src/wallet/wallet2.cpp#L836

Sorry I just typed off the name quickly, it was in `wallet2::get_subaddress_spend_public_key`. This is doing the same thing, but hasn't been removed.

And the speedup came from unpacking the public key once, otherwise its going to be really close. Guess its time to finally introduce `crypto::fast_public_key` (mymonero has been using this specific speedup for wallet scanning for a bit).

What I did was create a new type in `crypto.h` that was equal to the size of `ge_pe3`. Then did a memcpy internally in `crypto.cpp`. I think it was a bit more reusable since it kept the crypto in one spot - unpacking a key once for multiple `scalarmult`s is useful in other contexts. Although in my situation it was doing from an arbitrary base, so it wasn't the identical function.

Reply:
Let me re-clarify a bit, if the function is still being used elsewhere (I think it is), you can definitely merge the two paths easily. Theres no reason to alias the remainder of the function, its just an unnecessary duplication.